### DISCUSSION TITLE
draft proposal? support "non destructive mutation" on F# records

### DISCUSSION THREAD

**Discussion by smoothdeveloper** [dotnet/csharplang#6067] [Language Ideas]
Microsoft added records types to C#, and they support "non destructive mutation":

https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/records#non-destructive-mutation

in the same way F# supports it on records types:

https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/copy-and-update-record-expressions

I'd like C# to support the construct on F# records, which are plenty in dotnet codebases.

Bonus point if the support can leverage the code generation of the F# compiler in cases semantics would be different.

**Comment by svick** [comment]
Personally, I don't think F# records should be special cased here. Instead, there should be a general pattern that allows using non-destructive mutation on a type. That pattern should be made with F# in mind, but it's probably okay if F# has to make some small adjustment to make it work.

I have found [these Language Design Meeting notes](https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-06-22.md#with-expressions-on-non-records), which say this should be considered "after C# 9". Not sure what's the current status of that.

**Reply by smoothdeveloper** [reply]
> there should be a general pattern that allows using non-destructive mutation on a type

> That pattern should be made with F# in mind, but it's probably okay if F# has to make some small adjustment to make it work.

Agreed, and adjusting the F# compiler on those grounds is of course something to consider, I like the way you frame it with "non-destructive mutation" as more general concept and seen from C# standpoint.

Thanks for great feedback!

**Reply by smoothdeveloper** [reply]
@svick great find about the LDM notes, also, @dsyme, the `with` syntax or "constructing a clone with arbitrary adjustments" is kind of being considered for other types than the flavor of record types.

Could we frame something with C# team to have a first proof of concept of what it means, on F# defined records?

**Reply by HaloFour** [reply]
What the C# compiler emits today looks like this:

```cs
var point1 = new Point { X = 2, Y = 3);
var point2 = point1 with { X = 3 };

// generates
var $temp = point1.Clone();  // the name of this method is "unspeakable": <Clone>$
$temp.X = 3;
var point2 = $temp;
```

This is different from F# that uses the type metadata to figure out which property goes to which constructor parameter and does this instead:

```fsharp
let p2 = { p1 with X = 3 }

// generates, in C# form
var p2 = new Point(3, p1.Y);
```

The major reason that the clone is unspeakable today is because unless it explicitly virtual and implemented by each type within a hierarchy you could end up decapitating the original value:

```cs
Person person = new Student(...);
Person person2 = person with { ... };

Debug.Assert(person2 is Student); // oops, not a Student!
```

**Comment by HaloFour** [comment]
F# relies on custom metadata in the form of attributes to describe types like records and to denote how they work.  I don't think it would be appropriate for the C# language to understand how to interpret this metadata.

**Reply by smoothdeveloper** [reply]
> F# relies on custom metadata in the form of attributes to describe types like records and to denote how they work.

May be you also know how it works in C#? if you mind expanding on the technicals you may have at grasp.

I am sure this can be revised in the way F# compiler emits the current metadata, by putting it in a form that the C# compiler understands on top of what F# knows, for the time being, but that won't solve the mismatch for assemblies currently compiled.

Would you like to submit an issue on dotnet/fsharp if you have technical insights on the matter?

> I don't think it would be appropriate for the C# language to understand how to interpret this metadata.

Do you mind expanding on the reasons? Are they technical or otherwise?

I personally think the concept of records and how they use `with` syntax for cloning with adjustments of members is something that should work for record types in CLR codebases, disregarding which compiler it is, and it is not a hard technical problem, if it is relevant one to tackle.

Right now, this can probably be achieved, without breaking changes, and a bit of work on both compilers, so it works for types that they both emit and exist in the wild.

Thanks a lot for sharing your perspective!

**Reply by HaloFour** [reply]
> May be you also know how it works in C#? if you mind expanding on the technicals you may have at grasp.

As of right now the compiler relies on an unspeakable method that clones the record and temporarily allows the `init` members to be written.  The C# team has mentioned wanting to make that pattern based so that other types can participate (including from other languages).  F# could technically emit that unspeakable method today, but IMO it'd be better if both teams could come up with a pattern that allows developers in any language to define these kinds of types.

> Do you mind expanding on the reasons? Are they technical or otherwise?

Specifically that the C# language specification shouldn't reference attributes that are specific to F# and exist in `Microsoft.FSharp.Core`, such as `CompilationMapping` which F# uses to mark that a type is a record and to specify how the nominal properties relate to the positional constructor.  The F# compiler uses that both for nominal construction as well as to figure out how to wire the properties that are being copied as-is to the new record.

**Reply by smoothdeveloper** [reply]
Thanks a ton, amazing insights you have!

> but IMO it'd be better if both teams could come up with a pattern that allows developers in any language to define these kinds of types.

I share the same opinion, and probably, some of the infrastructure for a more synergistic and open design would be part of several compiler implementations.

> Specifically that the C# language specification shouldn't reference attributes that are specific to F# and exist in Microsoft.FSharp.Core, such as CompilationMapping which F# uses to mark that a type is a record and to specify how the nominal properties relate to the positional constructor. 

I'll inquire on dotnet/fsharp if there is ground for speaking the "unspeakable method that clones" that C# does speak, this much likely lays ground to infrastucture to handle the converse feature from F# standpoint (consuming the CLR record types that C# emits).

**Reply by HaloFour** [reply]
It's only that I've had this conversation before a couple of times, and I've also experienced the same with Java and Scala, especially as Java now adopts more functional-friendly features and Scala finds itself having to adapt as well.  I guess that's always the problem with being the smaller but arguably more innovative language.

**Reply by smoothdeveloper** [reply]
@HaloFour thanks, I believe the F# compiler can adapt, but I also see a potential where some aspects of code generation for several dotnet languages to basically cross polinate / cross consume for better good of the ecosystem.

To me, it would make sense both the F# and C# compiler leverages the code generation infrastructure of each others, to be sure that the semantics of the types carry on for the special syntax for the specific operations.

Reality that you state sounds like C# is defining "manipulating a record in CLR land" and F# and others blend in support for the fitting construct, tacitly, and we call the "C# got a feature" "dust settles in CLR land" from standpoint of non C# languages.

Ideally, for compilers implementation and infrastructure which may make the IL itself evolve later on, maybe code generation infrastructure that is cross consumed between several CLI compilers, would also facilitate some aspects of such transitioning.

It maybe another case where we want to "do better than Java" in this regard, it was one aspect behind C# language in particular, and also the dotnet runtime; and I believe, in the long run, more "compiler infrastructure that scales many languages with no N x M issues" will be key to dotnet success.

To me, it requires a layer underlying Roslyn and F# Compiler Service in context of our current discussion (if taken outside context of "let's records be friends in F# and C# languages" specific feature).

If we are building dotnet for C# mostly, then things are different of course.

**Reply by CyrusNajmabadi** [reply]
I'm not really sure what is being asked for.  Each language can decide waht it's desired experience is around the types generated by the other language :)

**Reply by HaloFour** [reply]
F# doesn't use Roslyn, it has its own compiler.  But even if it did my understanding is that the majority of the language support is not shared between languages and that was not one of its initial design goals.

**Reply by smoothdeveloper** [reply]
@CyrusNajmabadi thanks for the prompt and I agree with the philosophy if looked 100% outside of any context.

In the specifics, what would you say is the standpoint for C# philosophy to support the "non destructive mutation" semantics on record types that exists in assemblies compiled with say F# 1.0 or later?

It is what my request stands for.

On F# side, it is likely the support for C# defined records slowly surfacing in codebases can be used as if they were records as known in F# (in terms of particular syntax for constructing / destructing and other aspects too): I say likely because up for discussion in the details, and got discussed as "making sense" already in some other places: https://github.com/fsharp/fslang-suggestions/issues/1138

As someone using both language in different codebases, and libraries written in both, I do see value in asking for different things:

* record types in all existing dotnet assemblies, at large, could be used today in C#, with same support as if they are now possible to define in C#
* asking if there aren't parts related to code generation, that shouldn't go under both Roslyn and F#, conceptually, if we'd like to tackle this at lower level, for possibly other languages coming to dotnet and things related to evolution of IL

Same "me asking" will likely occur when C# gets support for defining DU or something akin to it, that it knows how to do the same thing with F# defined DU for the common use case people already use them in the dotnet eco system.

Anything the C# maintainers feel is good for dotnet codebases, I think should be considered.

I know it may fall outside "Each language can decide waht it's desired experience is around the types generated by the other language :)". 

There are places where it is not most conducive to people doing dotnet without having "the other language" in mind (because it would be plural at the very least, or simply wouldn't matter, in terms of dealing with real world code, which is just maintained over the years).

We should look at those things as "dotnet records" for a bit and make dotnet languages who deal with records nicely update themselves for good of compatibility and maintainability of codebases running on dotnet.

@HaloFour 

> But even if it did my understanding is that the majority of the language support is not shared between languages and that was not one of its initial design goals.

What is wrong at looking at code generation and the concepts that fit under "dotnet records" that both languages deal with in same manner (I gather C# took inspiration on F# records at some levels)?

It doesn't matter Roslyn compiler or F# compiler don't use each others, if we'd like to consider it would be a good thing to have a lower layer, in terms of moving state of dotnet compilers and dotnet codebases forward, without artificial barriers.

This technical aspect is not needed for both F# and C# to improve in handling existing assemblies having "dotnet records" defined in them, but it may be a nice PoC area if the teams working on those big pieces of dotnet, would consider it.

**Reply by HaloFour** [reply]
@smoothdeveloper 

> It doesn't matter Roslyn compiler or F# compiler don't use each others, if we'd like to consider it would be a good thing to have a lower layer, in terms of moving state of dotnet compilers and dotnet codebases forward, without artificial barriers.

That lower layer is the metadata in the assembly, which is simple enough to read using the APIs in the runtime.  If your request is for some new kind of abstraction library to exist underneath both Roslyn and F# compilers I don't think you'd find much support for that, nor would one be likely to help, as each feature needs to be interpreted differently by each language as they won't share the same grammar if they have similar features.
